File: RobWork/example/cpp/ex-invkin.cpp
#include <rw/invkin/ClosedFormIKSolverUR.hpp>
#include <rw/kinematics/State.hpp>
#include <rw/loaders/WorkCellLoader.hpp>
#include <rw/math/EAA.hpp>
#include <rw/math/Transform3D.hpp>
#include <rw/math/Vector3D.hpp>
#include <rw/math/Q.hpp>
#include <rw/models/SerialDevice.hpp>
#include <rw/models/WorkCell.hpp>

#include <iostream>

using rw::invkin::ClosedFormIKSolverUR;
using rw::kinematics::State;
using rw::loaders::WorkCellLoader;
using namespace rw::math;
using namespace rw::models;

#define WC_FILE "/devices/serialdev/UR10e_2018/UR10e.xml"

int main(int argc, char** argv) {
    if (argc != 2) {
        std::cout << "Usage: " << argv[0] << " <path/to/RobWorkData>" << std::endl;
        exit(1);
    }
    const std::string path = argv[1];

    const WorkCell::Ptr wc = WorkCellLoader::Factory::load(path + WC_FILE);
    if (wc.isNull())
        RW_THROW("WorkCell could not be loaded.");
    const SerialDevice::Ptr device = wc->findDevice<SerialDevice>("UR10e");
    if (device.isNull())
        RW_THROW("UR10e device could not be found.");

    const State state = wc->getDefaultState();
    const ClosedFormIKSolverUR solver(device, state);

    const Transform3D<> Tdesired(Vector3D<>(0.2, -0.2, 0.5),
            EAA<>(0, Pi, 0).toRotation3D());
    const std::vector<Q> solutions = solver.solve(Tdesired, state);

    std::cout << "Inverse Kinematics for " << device->getName() << "." << std::endl;
    std::cout << " Base frame: " << device->getBase()->getName() << std::endl;
    std::cout << " End/TCP frame: " << solver.getTCP()->getName() << std::endl;
    std::cout << " Target Transform: " << Tdesired << std::endl;
    std::cout << "Found " << solutions.size() << " solutions." << std::endl;
    for(std::size_t i = 0; i < solutions.size(); i++) {
        std::cout << " " << solutions[i] << std::endl;
    }

    return 0;
}

File: RobWork/example/lua/ex-invkin.lua
require("sdurw_core")
require("sdurw_math")
require("sdurw_kinematics")
require("sdurw_models")
require("sdurw")

using("sdurw_math")
using("sdurw_kinematics")
using("sdurw_models")
using("sdurw")

if #arg ~= 1 then
    print("Usage: lua ex-invkin.lua <path/to/RobWorkData>")
    return 1
end

local WC_FILE = "/devices/serialdev/UR10e_2018/UR10e.xml"

local wc = WorkCellLoaderFactory.load(arg[1] .. WC_FILE)
if wc:isNull() then
    error("WorkCell could not be loaded")
end
local device = wc:findSerialDevice("UR10e")
if device:isNull() then
    error("UR10e device could not be found.")
end

local state = wc:getDefaultState()
local solver = ClosedFormIKSolverUR(device:cptr(), state)

local Tdesired = Transform3Dd(Vector3Dd(0.2, -0.2, 0.5), EAAd(0, Pi, 0):toRotation3D())
local solutions = solver:solve(Tdesired, state)

print("Inverse Kinematics for " .. device:getName() .. ".")
print(" Base frame: " .. device:getBase():getName())
print(" End/TCP frame: " .. solver:getTCP():getName())
print(" Target Transform: " .. tostring(Tdesired))
print("Found " .. solutions:size() .. " solutions.")
for i= 0,solutions:size()-1,1
do
    print(" " .. tostring(solutions[i]))
end
